### ISSUE TITLE
Support multi-threading for large images on Windows.

### ISSUE THREAD

**Issue by ufechner7** [JuliaLang/julia#52561]
MWE:
Check out from git:
```
cd repos # any folder of your choice, but without spaces in the folder name
git clone https://github.com/ufechner7/Tethers.jl
```
Build the system image:
```
cd repos/Tethers.jl
cd bin
./create_sys_image
```
Now look into the task manager and check the CPU load. My CPU load was mostly at 7%, sometimes a bit higher,
but 8 threads where not used when creating the system image.

Doing the same on Linux is much faster, and the CPU load reaches 800% after about 2 minutes of the progress
of "building system image".

I used `juliaup` to run Julia. The machine had 24GB of RAM, never more than 50% where in use.
```
julia> versioninfo()
Julia Version 1.10.0-rc2
Commit dbb9c46795 (2023-12-03 15:25 UTC)
Build Info:
  Official https://julialang.org/ release
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
  CPU: 16 × AMD Ryzen 9 7950X 16-Core Processor
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-15.0.7 (ORCJIT, znver3)
  Threads: 1 on 16 virtual cores
```
Summary: Building a system image is slow on Windows and fast on Linux because on Windows Julia 1.10-rc2 uses only one thread.

Is this a known limitation?

---
How it should work is documented: 
[JULIA_IMAGE_THREADS](https://docs.julialang.org/en/v1.10.0-rc2/manual/environment-variables/#env-image-threads)

An unsigned 32-bit integer that sets the number of threads used by image compilation in this Julia process. The value of this variable may be ignored if the module is a small module. If left unspecified, the smaller of the value of [JULIA_CPU_THREADS](https://docs.julialang.org/en/v1.10.0-rc2/manual/environment-variables/#env-cpu-threads) or half the number of logical CPU cores is used in its place.

**Comment by vtjnash** [comment]
We discovered that Windows did not behave itself well, so we disabled it, especially as very little development is actually done directly on that platform

**Comment by ufechner7** [comment]
That is a pity... I want to use ModelingToolkit in a class, and I have to suggest the students to create a system image because loading it is still slow, even with Julia 1.10. It makes a difference if that takes 15 min or 30 min, though... 
Any chance to re-enable it at some point in time?

**Comment by vchuravy** [comment]
>  Any chance to re-enable it at some point in time?

I don't think we hard disable parallelism per-se, but ironically only when the system images are very large:
https://github.com/JuliaLang/julia/blob/b57f8d168dc88be8570e1c9ff7cd352fc5e11c98/src/aotcompile.cpp#L1454

x-ref: #50874

At the core of the issue is that COFF (the default binary format on Windows) uses 16bit for the number of symbols.

**Comment by ufechner7** [comment]
How do you calculate what is "very large"?

**Comment by KristofferC** [comment]
I think it is https://github.com/JuliaLang/julia/blob/590a63b299880e874ca589fb8c9a6f7b771cd701/src/aotcompile.cpp#L1420

**Comment by ufechner7** [comment]
Could we use the https://en.wikipedia.org/wiki/Portable_Executable format on Windows?

**Comment by gbaraldi** [comment]
PE is pretty much COFF, COFF was the unix name and PE is the microsoft name

**Comment by ufechner7** [comment]
> PE is pretty much COFF, COFF was the unix name and PE is the microsoft name

Maybe a version with 64-bit address space support called PE32+ ?

**Comment by gbaraldi** [comment]
It's going to be the same, the issue is that they use 16bit integer do identify stuff, and that has a 65k maximum value. People have complained to microsoft lots of times about this, but they haven't really fixed this

**Comment by ufechner7** [comment]
Or can we reduce the number of exported symbols?

**Comment by vchuravy** [comment]
When we use multiple threads we split the output into multiple files itself, but you still want calls across the split files to be fast.
So we export the symbols in the split files and then we merge them into one final output file, which then fails since we have to many symbols.

In single-threaded compilation we never make the symbols visible.

**Comment by ufechner7** [comment]
This sounds to me as if we could use another file format for the split files...

**Comment by vchuravy** [comment]
To my knowledge that is not feasible, but I would be happy to be proven wrong.
Windows is not a platform that we have many (or really any developers for).

**Comment by ufechner7** [comment]
> To my knowledge that is not feasible

Why not? Which program produces the intermediate files? And which program combines them?

**Comment by ufechner7** [comment]
> Windows is not a platform that we have many (or really any developers for).

Perhaps you should ask Microsoft for a developer to help out? They must be interested that Windows user have a good experience when using Julia...

**Comment by vchuravy** [comment]
Thanks Prem for the like. To summarize my understanding Microsoft added `COFF+/bigobj` many years ago that allows for more than 16k symbols, but failed to document it sufficiently.
This means open-source tools chains (like ours) build on top of LLVM struggle with using `\bigobj` (looks.like LLVM can maybe detect/read it but not write it?)

**Comment by awson** [comment]
LLVM **tools** handle bigobj perfectly, you don't even need to feed them with this option (e.g. clang automatically determines if it needs to generate bigobj file).

PE number of exported symbols limit is completely different story, it isn't related to bigobj story and, I believe, will never be fixed by MS. Thus, to overcome this limit you shall invent your own exporting (and loading) scheme. E.g., I had developed such a scheme scheme in my [GHC on native Windows SDK project](https://awson.github.io/ghc-nw/).

I don't know which problem you face here, but if it is the former (bigobj) it should be easily fixable.

**Comment by vchuravy** [comment]
No I think you are right I was looking at this yesterday and bigobj is related to sections and we are facing number of symbols here.

**Comment by vchuravy** [comment]
This is the original error:

**Comment by awson** [comment]
> When we use multiple threads we split the output into multiple files itself, but you still want calls across the split files to be fast. So we export the symbols in the split files and then we merge them into one final output file, which then fails since we have to many symbols.
> 
> In single-threaded compilation we never make the symbols visible.

I guess this is a mistake.

**Comment by pchintalapudi** [comment]
They're not dllexport, only extern hidden (global aliases are a rarer case that we do need to dllexport for language reasons). We mark the symbols with hidden visibility on line 838, and we also confirmed that they're hidden in the final shared object on Linux.

**Comment by awson** [comment]
> They're not dllexport, only extern hidden (global aliases are a rarer case that we do need to dllexport for language reasons). We mark the symbols with hidden visibility on line 838, and we also confirmed that they're hidden in the final shared object on Linux.

If they aren't dllexport the linker shouldn't try to export them then, unless it is invoked with `--export-all-symbols` or something like this.
